### PR TITLE
Add support for large models

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Current implementation exports pytorch models to onnx including initializers, which are moved to input nodes and deleted from initializer list. Because initializers are included, 2GB size limit is easily reached by big models.

This PR exports the model without initializers, allowing large models to be exported and trained

### CODE REVIEW

**Comment by mrry** [comment]
Looks like it works for everything apart from BERT:

```
E               RuntimeError: Input is present in ONNX graph but not provided: _base_module.bert.embeddings.position_ids.
```

...perhaps because it's handled via `self.register_buffer()` instead of the regular setattr approach?

https://github.com/huggingface/transformers/blob/a735f727cc93f96abc5b39e776716891667bb565/src/transformers/models/bert/modeling_bert.py#L181

At a guess, do we need to be using `self.named_buffers()` instead of `self.named_parameters()` when passing the inputs corresponding to initializers?

File: orttraining/orttraining/python/training/ortmodule.py [hunk 1]
```diff
@@ -443,6 +439,12 @@ def _convert_training_graph_input_to_list(self, *inputs, **kwargs):
                 inp = non_none_inputs[input_idx]
             elif name in kwargs and kwargs[name] is not None:
                 inp = kwargs[name]
+            elif input_idx >= len(non_none_inputs):
+                # Registered buffers are translated to user_input+initializer in ONNX
+                buffer_name, inp = next(self._original_module.named_buffers())
```
Comment (mrry) [inline orttraining/orttraining/python/training/ortmodule.py:444]: IIUC this is always taking the *first* element from `self._original_module.named_buffers()`.

What happens if there are two or more registered buffers? Can you add unit tests for these cases?

**Review by mrry — APPROVED** [review]
Thanks Thiago! Looks good to me :).

File: orttraining/orttraining/python/training/ortmodule.py [hunk 2]
```diff
@@ -436,13 +432,20 @@ def _convert_training_graph_input_to_list(self, *inputs, **kwargs):
         '''
         # User inputs
         non_none_inputs = [inp for inp in inputs if inp is not None]
+        named_buffers_iter = iter(self._original_module.named_buffers())
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule.py:435]: shall we use the ```_flattened_output_module``` here instead?

File: orttraining/orttraining/python/training/ortmodule.py [hunk 3]
```diff
@@ -436,13 +432,20 @@ def _convert_training_graph_input_to_list(self, *inputs, **kwargs):
...
             inp = None
             if input_idx < len(non_none_inputs):
                 inp = non_none_inputs[input_idx]
             elif name in kwargs and kwargs[name] is not None:
                 inp = kwargs[name]
+            elif input_idx >= len(non_none_inputs):
+                # Registered buffers are translated to user_input+initializer in ONNX
+                buffer_name, inp = next(named_buffers_iter)
+                # `name` contains a `_base_module` prefix from `FlattenedOutputModule`
+                assert buffer_name in name, f'Input name {name} expected, but {buffer_name} found!'
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule.py:447]: It is also possible here that the user changed the number of inputs from the earlier call to the model to this one. In that scenario, this assert will fail, right? Shall we add a TODO to handle the case when the number of arguments to forward changes from one call to another (in which case, we will have to re-export)?

Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule.py]: Added a TODO and will make sure UTs for this scenario is pushed on a seParate PR

**Review by baijumeswani — APPROVED** [review]
Looks good.

**Pull request merged**